server/drivers/linux_devlcd.c

```diff
@@ -0,0 +1,312 @@
+	}
+	else {
+		/* Use our own size from config file */
+		strncpy(buf, drvthis->config_get_string(drvthis->name, "Size", 0, TEXTDRV_DEFAULT_SIZE), sizeof(buf));
+		buf[sizeof(buf)-1] = '\0';
+		if ((sscanf(buf , "%dx%d", &p->width, &p->height) != 2)
+		    || (p->width <= 0) || (p->width > LCD_MAX_WIDTH)
+		    || (p->height <= 0) || (p->height > LCD_MAX_HEIGHT)) {
+			report(RPT_WARNING, "%s: cannot read Size: %s; using default %s",
+					drvthis->name, buf, TEXTDRV_DEFAULT_SIZE);
+			sscanf(TEXTDRV_DEFAULT_SIZE, "%dx%d", &p->width, &p->height);
+		}
+	}
+
+	// Allocate the framebuffer
+	p->framebuf = malloc(p->width * p->height);
+	if (p->framebuf == NULL) {
+		report(RPT_ERR, "%s: unable to create framebuffer", drvthis->name);
+		return -1;
+	}
+	memset(p->framebuf, ' ', p->width * p->height);
+
+	/* open the device... */
+	if (!strcmp(device,"-")) {
+		p->fd = stdout;
+	} else {
+		p->fd = fopen(device, "w" );
+	}
+	if (p->fd == 0) {
+		report(RPT_ERR, "%s: open(%s) failed (%s)", drvthis->name, device, strerror(errno));
+		if (errno == EACCES)
+			report(RPT_ERR, "%s: device %s could not be opened", drvthis->name, device);
+		goto err_out;
+	}
+	report(RPT_INFO, "%s: opened display on %s", drvthis->name, device);
+
+	fprintf(p->fd, "\e[LI"); // Reinitialise display
+
+	report(RPT_DEBUG, "%s: init() done", drvthis->name);
+
+	return 0;
+
+      err_out:
+	linuxDevLcd_close(drvthis);
+	return -1;
+}
+
+
+/**
+ * Close the driver (do necessary clean-up).
+ * \param drvthis  Pointer to driver structure.
+ */
+MODULE_EXPORT void
+linuxDevLcd_close (Driver *drvthis)
+{
+	PrivateData *p = drvthis->private_data;
+
+	if (p != NULL) {
+		if (p->framebuf != NULL)
+			free(p->framebuf);
+
+		free(p);
+	}
+	drvthis->store_private_ptr(drvthis, NULL);
+}
+
+
+/**
+ * Return the display width in characters.
+ * \param drvthis  Pointer to driver structure.
+ * \return         Number of characters the display is wide.
+ */
+MODULE_EXPORT int
+linuxDevLcd_width (Driver *drvthis)
+{
+	PrivateData *p = drvthis->private_data;
+
+	return p->width;
+}
+
+
+/**
+ * Return the display height in characters.
+ * \param drvthis  Pointer to driver structure.
+ * \return         Number of characters the display is high.
+ */
+MODULE_EXPORT int
+linuxDevLcd_height (Driver *drvthis)
+{
+	PrivateData *p = drvthis->private_data;
+
+	return p->height;
+}
+
+
+/**
+ * Clear the screen.
+ * \param drvthis  Pointer to driver structure.
+ */
+MODULE_EXPORT void
+linuxDevLcd_clear (Driver *drvthis)
+{
+	PrivateData *p = drvthis->private_data;
+
+	memset(p->framebuf, ' ', p->width * p->height);
+}
+
+
+/**
+ * Flush data on screen to the display.
+ * \param drvthis  Pointer to driver structure.
+ */
+MODULE_EXPORT void
+linuxDevLcd_flush (Driver *drvthis)
+{
+	PrivateData *p = drvthis->private_data;
+	char out[LCD_MAX_WIDTH];
+	int i;
+
+	fprintf(p->fd, "\e[H"); // cursor home
+	for (i = 0; i < p->height; i++) {
+		memcpy(out, p->framebuf + (i * p->width), p->width);
```

Comment:
Do we need some check for escape sequences here?

Reply:
Do you mean to filter out a user incorporating escape sequences in their strings causing odd effects ?
If so, the quickest (and dirtiest) way would be to drop any \e characters we see.
The more complete way is to re-implement the kernel's parser but that could be fiddly and need ongoing maintenance.
(If the kernel required the escape sequences to be ; terminated it would be much easier, but only one of them is ; terminated)

Reply:
If \e is the only character, that might cause trouble, then just replacing \e with some other inoffensive character would be the way to go.

Again this seems to boil down to reading the kernel code. Which is kind of bad because the kernel code might change without us noticing. Such things should be specified somewhere. But it's not your fault. As long as we do reasonable guarding on our end, we are good.